### PR TITLE
UCP/UCT: add a new env parameter to control the listener's backlog

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -291,6 +291,10 @@ static ucs_config_field_t ucp_config_table[] = {
    "require out of band synchronization before destroying UCP resources.",
    ucs_offsetof(ucp_config_t, ctx.sockaddr_cm_enable), UCS_CONFIG_TYPE_TERNARY},
 
+  {"LISTENER_BACKLOG", "auto",
+   "Maximum number of pending connections for a listener.",
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_context.c:295]: connections -> connection requests

File: src/ucp/core/ucp_context.h [hunk 2]
```diff
@@ -100,6 +100,8 @@ typedef struct ucp_context_config {
     int                                    unified_mode;
     /** Enable cm wireup-and-close protocol for client-server connections */
     ucs_ternary_value_t                    sockaddr_cm_enable;
+    /** Maximum number of pending connections for a listener */
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_context.h:103]: the same

File: src/uct/api/uct.h [hunk 3]
```diff
@@ -1205,7 +1205,7 @@ struct uct_listener_params {
 
     /**
      * Backlog of incoming connection requests.
-     * If not specified, SOMAXCONN, as defined in <sys/socket.h>, will be used.
+     * If not specified, each CM transport will select its own value.
```
Comment (evgeny-leksikov) [inline src/uct/api/uct.h:1208]: each CM component

File: src/uct/base/uct_cm.c [hunk 4]
```diff
@@ -111,6 +111,23 @@ void uct_cm_ep_server_conn_notify_cb(uct_cm_base_ep_t *cep, ucs_status_t status)
...
+    int backlog;
+
+    if ((params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) &&
+        (params->backlog == INT_MAX)) {
+        backlog = default_value;
+    } else if (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) {
+        backlog = params->backlog;
+    } else {
+        backlog = default_value;
+    }
```
Comment (evgeny-leksikov) [inline src/uct/base/uct_cm.c:126]: then it should be:
```
    if (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) {
        if (params->backlog > max_value) {
            ucs_diag(...);
        }
        backlog = ucs_min(params->backlog, max_value);
    } else {
        backlog = max_value;
    }

    if (backlog == 0) {
        ucs_error(...);
    }
```

File: src/uct/base/uct_cm.c [hunk 5]
```diff
@@ -111,6 +111,23 @@ void uct_cm_ep_server_conn_notify_cb(uct_cm_base_ep_t *cep, ucs_status_t status)
     cep->server.notify_cb(&cep->super.super, cep->user_data, &notify_args);
 }
 
+int uct_cm_set_listener_backlog(const uct_listener_params_t *params,
+                                int default_value)
```
Comment (evgeny-leksikov) [inline src/uct/base/uct_cm.c:115]: rename to `unsigned uct_listener_backlog_adjust(const uct_listener_params_t *params, unsigned max_value)` since it does not set anything
also `default_value` is actually `max_value` since you detect it explicitly for rdma_cm but for TCP it's:
"       If   the   backlog   argument   is    **greater**    than    the    value    in
       /proc/sys/net/core/somaxconn,  then it is **silently truncated to that value**;
       the default value in this file is 128.   In  kernels  before  2.4.25,  this
       limit was a hard coded value, SOMAXCONN, with the value 128.
" (c) man listen

File: src/uct/ib/rdmacm/rdmacm_listener.c [hunk 6]
```diff
@@ -41,8 +60,8 @@ UCS_CLASS_INIT_FUNC(uct_rdmacm_listener_t, uct_cm_h cm,
         goto err_destroy_id;
     }
 
-    backlog = (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) ?
-              params->backlog : SOMAXCONN;
+    backlog = uct_cm_set_listener_backlog(params, ucs_rdmacm_max_backlog());
```
Comment (evgeny-leksikov) [inline src/uct/ib/rdmacm/rdmacm_listener.c:63]: it should be out parameter of `uct_listener_query`

File: src/uct/ib/rdmacm/rdmacm_listener.c [hunk 7]
```diff
@@ -11,6 +11,25 @@
...
+static long ucs_rdmacm_max_backlog()
+{
+    long max_backlog;
+
+    if (ucs_read_file_number(&max_backlog, 1,
+                             UCS_RDMACM_MAX_BACKLOG_PATH) == UCS_OK) {
+        ucs_assert(max_backlog <= INT_MAX);
+        return max_backlog;
+    } else {
+        ucs_warn("unable to read max_backlog value from %s file",
```
Comment (evgeny-leksikov) [inline src/uct/ib/rdmacm/rdmacm_listener.c:26]: is rdmacm able to work without this file?

File: src/uct/base/uct_cm.c (hunk 5 above)
Comment (alinask) [reply src/uct/base/uct_cm.c:115]: thanks for this comment! 
actually in kernel 3.10 it's already a greater value:
```
alinas@jazz04 ~
$cat /proc/sys/net/core/somaxconn
15000
alinas@jazz04 ~
$uname -r
3.10.0-957.el7.x86_64
```

so this means that with the new env param, we can only decrease the backlog value for tcp.

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 6 above)
Comment (alinask) [reply src/uct/ib/rdmacm/rdmacm_listener.c:63]: @yosefe should we add the backlog to uct_listener_attr's api?

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 7 above)
Comment (alinask) [reply src/uct/ib/rdmacm/rdmacm_listener.c:26]: i don't know. but this warning will printed if this happens.

File: src/uct/base/uct_cm.c (hunk 4 above)
Comment (alinask) [reply src/uct/base/uct_cm.c:126]: this means to cut the value for rdmacm as well.

File: src/ucp/core/ucp_listener.c [hunk 8]
```diff
@@ -242,9 +242,11 @@ ucp_listen_on_cm(ucp_listener_h listener, const ucp_listener_params_t *params)
     ucs_assert_always(num_cms > 0);
 
     uct_params.field_mask       = UCT_LISTENER_PARAM_FIELD_CONN_REQUEST_CB |
-                                  UCT_LISTENER_PARAM_FIELD_USER_DATA;
+                                  UCT_LISTENER_PARAM_FIELD_USER_DATA |
```
Comment (dmitrygx) [inline src/ucp/core/ucp_listener.c:245]: minor: align by `|`

File: src/ucp/core/ucp_listener.c [hunk 9]
```diff
@@ -242,9 +242,11 @@ ucp_listen_on_cm(ucp_listener_h listener, const ucp_listener_params_t *params)
...
+                                  UCT_LISTENER_PARAM_FIELD_BACKLOG;
     uct_params.conn_request_cb  = ucp_cm_server_conn_request_cb;
     uct_params.user_data        = listener;
+    uct_params.backlog          = ucs_min(INT_MAX, worker->context->config.ext.listener_backlog);
```
Comment (dmitrygx) [inline src/ucp/core/ucp_listener.c:249]: maybe use `-1` for `UCS_ULUNITS_AUTO`.
otherwise, it is not obvious, when `UCX_LISTENER_BACKLOG=inf` resulted to the default value for backlog instead of the maximal possible value (i.e. INT_MAX).

File: src/uct/base/uct_cm.c [hunk 10]
```diff
@@ -111,6 +111,23 @@ void uct_cm_ep_server_conn_notify_cb(uct_cm_base_ep_t *cep, ucs_status_t status)
...
+{
+    int backlog;
+
+    if ((params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) &&
+        (params->backlog == INT_MAX)) {
```
Comment (dmitrygx) [inline src/uct/base/uct_cm.c:120]: if uses `-1` for backlog as a `AUTO`, fix this place

File: src/uct/ib/rdmacm/rdmacm_listener.c [hunk 11]
```diff
@@ -11,6 +11,25 @@
 #include "rdmacm_listener.h"
 
 
+#define UCS_RDMACM_MAX_BACKLOG_PATH        "/proc/sys/net/rdma_ucm/max_backlog"
+
+
+static long ucs_rdmacm_max_backlog()
+{
+    long max_backlog;
```
Comment (dmitrygx) [inline src/uct/ib/rdmacm/rdmacm_listener.c:19]:
```
{
    static long max_backlog = 0;

    if (max_backlog != 0) {
        return max_backlog 
    }

    ...
}
```
to not read the file each time when this function is called, i.e. use the same approach as in `ucs_socket_max_conn()`

**Review by dmitrygx — CHANGES_REQUESTED** [review]

File: src/ucp/core/ucp_listener.c (hunk 9 above)
Comment (alinask) [reply src/ucp/core/ucp_listener.c:249]: backlog is an int (from uct's api) so i can't assign an unsigned long to an int..

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 11 above)
Comment (alinask) [reply src/uct/ib/rdmacm/rdmacm_listener.c:19]: you mean in case there is more than 1 rdmacm_listener?

File: src/uct/base/uct_cm.c (hunk 4 above)
Comment (evgeny-leksikov) [reply src/uct/base/uct_cm.c:126]: do you think rdmacm can support greater backlog than it's defined in `/proc/sys/net/rdma_ucm/max_backlog`?
My guessing is rdmacm behavior is similar to TCP, need to clarify this.

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 11 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_listener.c:19]: on other hand, read the file every time could adjust the parameter in runtime for new listeners without restart server process.

File: src/uct/base/uct_cm.c (hunk 5 above)
Comment (evgeny-leksikov) [reply src/uct/base/uct_cm.c:115]: seems like that and the only way to increase the backlog is system limits. (in case if UCX default values = max values)

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 6 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_listener.c:63]: i don't think we need in this PR, it's just internal value.

File: src/ucp/core/ucp_listener.c (hunk 9 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_listener.c:249]: > backlog is an int (from uct's api) so i can't assign an unsigned long to an int..

But it is ULUNITS right now, so user could specify `inf`, this will be `-1` when cast to `int`, right?

Comment (alinask) [reply src/ucp/core/ucp_listener.c:249]: but i'm not casting to int. if the user sets 'inf', since it's an unsigned long, it will be a very large number and the ucs_min between this number and INT_MAX will be INT_MAX which would be set into the int backlog.

Comment (dmitrygx) [reply src/ucp/core/ucp_listener.c:249]: ah, I see
I'm okay 👍 
But if this is`auto`, it will be `INT_MAX`, right? I'd expect to make it equal to the system default value in this case

Comment (alinask) [reply src/ucp/core/ucp_listener.c:249]: right, backlog would be set to INT_MAX here and then, inside uct, will be set to the system's setting.

Comment (dmitrygx) [reply src/ucp/core/ucp_listener.c:249]: ok, but if this is `inf`, it also would be adjusted to the system default, right?

**Review by dmitrygx — APPROVED** [review]

File: src/uct/base/uct_cm.c [hunk 12]
```diff
@@ -223,6 +223,28 @@ ucs_status_t uct_listener_reject(uct_listener_h listener,
...
+        if (params->backlog > max_value) {
+            ucs_diag("configure value %d is greater than the max_value %d. "
+                     "using max_value", params->backlog, max_value);
+        }
+        backlog = ucs_min(params->backlog, max_value);
+    } else {
+        backlog = max_value;
+    }
+
+    if (backlog == 0) {
```
Comment (evgeny-leksikov) [inline src/uct/base/uct_cm.c:241]: it should be `<= 0` since you are staying with `int`

File: src/uct/tcp/tcp_listener.c [hunk 13]
```diff
@@ -106,8 +106,12 @@ UCS_CLASS_INIT_FUNC(uct_tcp_listener_t, uct_cm_h cm,
     self->conn_request_cb = params->conn_request_cb;
     self->user_data       = (params->field_mask & UCT_LISTENER_PARAM_FIELD_USER_DATA) ?
                             params->user_data : NULL;
-    backlog               = (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) ?
-                            params->backlog : ucs_socket_max_conn();
+
+    backlog = uct_listener_backlog_adjust(params, ucs_socket_max_conn());
+    if (backlog == 0) {
```
Comment (evgeny-leksikov) [inline src/uct/tcp/tcp_listener.c:111]: the same

**Review by evgeny-leksikov — COMMENTED** [review]
IMHO: better to change internal representation of backlog from `int` to `unsigned` since we cut it down. this way no need to check for negative value.

**Comment by alinask** [comment]
@evgeny-leksikov both listen() and rdma_listen() accept backlog as an int. I think uct should be consistent with them.

**Comment by evgeny-leksikov** [comment]
> @evgeny-leksikov both listen() and rdma_listen() accept backlog as an int. I think uct should be consistent with them.

ok, then need to check negative value

File: src/uct/base/uct_cm.c (hunk 12 above)
Comment (alinask) [reply src/uct/base/uct_cm.c:241]: even though it can't really have a negative value?
`uct_params.backlog = ucs_min(INT_MAX, worker->context->config.ext.listener_backlog);`

Comment (evgeny-leksikov) [reply src/uct/base/uct_cm.c:241]: UCP is not the only UCT user, also it can be wrongly passed to UCP config

**Review by evgeny-leksikov — APPROVED** [review]

**Comment by alinask** [comment]
@yosefe Can you please review?

File: src/ucp/core/ucp_context.c [hunk 14]
```diff
@@ -291,6 +291,13 @@ static ucs_config_field_t ucp_config_table[] = {
    "require out of band synchronization before destroying UCP resources.",
    ucs_offsetof(ucp_config_t, ctx.sockaddr_cm_enable), UCS_CONFIG_TYPE_TERNARY},
 
+  {"LISTENER_BACKLOG", "auto",
+   "Maximum number of pending connection requests for a listener.\n"
+   "auto means that each transport would use its system default.\n"
+   "If a value larger than a system default is set, the backlog value would be\n"
+   "cut to the system default.",
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:298]:
```
'auto' means that each transport would use its maximal allowed value.
If a value larger than a transport supports is set, the backlog value would be
cut to that maximal value.
```

File: src/ucp/core/ucp_context.h [hunk 15]
```diff
@@ -100,6 +100,8 @@ typedef struct ucp_context_config {
     int                                    unified_mode;
     /** Enable cm wireup-and-close protocol for client-server connections */
     ucs_ternary_value_t                    sockaddr_cm_enable;
+    /** Maximum number of pending connection requests for a listener */
```
Comment (yosefe) [inline src/ucp/core/ucp_context.h:103]: Maximal

File: src/ucp/core/ucp_listener.c (hunk 9 above)
Comment (yosefe) [reply src/ucp/core/ucp_listener.c:249]: to make sure ucs_min is done right:
`ucs_min((size_t)INT_MAX, worker->context->config.ext.listener_backlog)`;

File: src/uct/base/uct_cm.c [hunk 16]
```diff
@@ -223,6 +223,28 @@ ucs_status_t uct_listener_reject(uct_listener_h listener,
...
+                                int max_value)
+{
+    int backlog;
+
+    if (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) {
+        if (params->backlog > max_value) {
+            ucs_diag("configure value %d is greater than the max_value %d. "
+                     "using max_value", params->backlog, max_value);
+        }
+        backlog = ucs_min(params->backlog, max_value);
```
Comment (yosefe) [inline src/uct/base/uct_cm.c:236]:
```suggestion
        if (params->backlog > max_value) {
            ucs_diag("configure value %d is greater than the max_value %d. "
                     "using max_value", params->backlog, max_value);
            backlog = max_value;
        } else {
            backlog = params->backlog;
        }
```

File: src/uct/base/uct_cm.c [hunk 17]
```diff
@@ -223,6 +223,28 @@ ucs_status_t uct_listener_reject(uct_listener_h listener,
...
+    } else {
+        backlog = max_value;
+    }
+
+    if (backlog <= 0) {
+        ucs_error("the backlog value cannot be zero or negative");
+    }
```
Comment (yosefe) [inline src/uct/base/uct_cm.c:243]: need to check backlog<=0 only once during the flow (either inside or outside the function)
if checking inside the func, return status which indicates OK/INVALID_PARAM

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 7 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_listener.c:26]: let's make it ucs_diag

File: src/uct/ib/rdmacm/rdmacm_listener.c (hunk 11 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_listener.c:19]: i think better read only once, we are doing same way in ucs_socket_max_conn()

File: src/uct/ib/rdmacm/rdmacm_listener.c [hunk 18]
```diff
@@ -11,6 +11,25 @@
...
+
+    if (ucs_read_file_number(&max_backlog, 1,
+                             UCS_RDMACM_MAX_BACKLOG_PATH) == UCS_OK) {
+        ucs_assert(max_backlog <= INT_MAX);
+        return max_backlog;
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_listener.c:24]: "return max_backlog" can be outside of if/else

File: src/uct/api/uct.h [hunk 19]
```diff
@@ -1205,7 +1205,7 @@ struct uct_listener_params {
 
     /**
      * Backlog of incoming connection requests.
-     * If not specified, SOMAXCONN, as defined in <sys/socket.h>, will be used.
+     * If not specified, each CM component will select its own value.
```
Comment (yosefe) [inline src/uct/api/uct.h:1208]: can we move the doc change to another PR?

Comment (alinask) [reply src/uct/api/uct.h:1208]: well we are changing the behavior of setting the backlog. for rdmacm, for example, the previous doc won't be correct any more...

**Comment by alinask** [comment]
@yosefe Can you take a look please?

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@tonycurtis @shamisp can you pls take a look, there is one line doc change in src/uct/api/uct.h

**Comment by yosefe** [comment]
> @tonycurtis @shamisp can you pls take a look, there is one line doc change in src/uct/api/uct.h

nevermind - it will be in a separate PR
